File: python/app/services/ai_service.py
import json
import random
from typing import Any

import httpx

from app.config import settings

AI_API_URL = settings.AI_API_URL
AI_API_KEY = settings.AI_API_KEY
AI_MODEL = settings.AI_MODEL

SYSTEM_PROMPTS = {
    "classify": "Você é um classificador de gastos. Analise a descrição e determine a categoria, tipo (receita/despesa) e tags relevantes. Responda APENAS com JSON.",
    "analyze_spending": "Você é um analista financeiro. Analise os gastos do usuário e forneça insights sobre padrões de consumo.",
    "anomalies": "Você é um analista financeiro. Detecte transações anômalas ou suspeitas nos dados fornecidos.",
    "predict_budget": "Você é um consultor financeiro. Com base no histórico de gastos, sugira um orçamento para o próximo mês.",
    "tips": "Você é um consultor financeiro. Forneça dicas personalizadas de economia com base nos hábitos de gastos do usuário.",
    "chat": "Você é um assistente financeiro pessoal. Responda perguntas sobre finanças e ajude o usuário a gerenciar seus gastos.",
}


async def _call_ai_api(messages: list[dict], temperature: float = 0.7) -> dict[str, Any] | None:
    if not AI_API_KEY or AI_API_KEY == "sua-chave-aqui":
        return None

    async with httpx.AsyncClient(timeout=30.0) as client:
        response = await client.post(
            f"{AI_API_URL}/chat/completions",
            headers={
                "Authorization": f"Bearer {AI_API_KEY}",
                "Content-Type": "application/json",
            },
            json={
                "model": AI_MODEL,
                "messages": messages,
                "temperature": temperature,
            },
        )
        response.raise_for_status()
        data = response.json()
        content = data["choices"][0]["message"]["content"]
        try:
            return json.loads(content)
        except (json.JSONDecodeError, KeyError, IndexError):
            return {"response": content}


def _fallback_classify(description: str) -> dict:
    categories = {
        "alimentação": ["restaurante", "mercado", "supermercado", "comida", "lanche", "pizza", " ifood", "shopper", "compras"],
        "transporte": ["uber", "99", "taxi", "ônibus", "metrô", "gasolina", "combustível", "pedágio"],
        "moradia": ["aluguel", "condomínio", "água", "luz", "energia", "gás", "internet"],
        "saúde": ["farmácia", "médico", "dentista", "plano de saúde", "hospital", "exame"],
        "educação": ["curso", "faculdade", "escola", "livro", "material"],
        "lazer": ["cinema", "show", "viagem", "jogo", "streaming", "netflix", "spotify"],
        "salário": ["salário", "pagamento", "recebimento", "depósito", "freela", "pro labore"],
        "investimentos": ["investimento", "ações", "renda fixa", "cdb", "tesouro"],
    }

    desc_lower = description.lower()
    for cat, keywords in categories.items():
        for kw in keywords:
            if kw in desc_lower:
                is_receita = cat in ("salário", "investimentos")
                return {
                    "category": cat,
                    "type": "receita" if is_receita else "despesa",
                    "tags": [kw],
                }

    return {
        "category": "outros",
        "type": "despesa",
        "tags": [],
    }


def _fallback_analyze_spending(transactions: list[dict]) -> dict:
    total = sum(t.get("amount", 0) for t in transactions)
    count = len(transactions)
    return {
        "insights": [
            f"Você registrou {count} transações no período.",
            f"Total movimentado: R$ {total:.2f}",
            "Continue monitorando seus gastos para melhores resultados financeiros.",
        ],
        "summary": {
            "total": round(total, 2),
            "count": count,
        },
    }


def _fallback_anomalies(transactions: list[dict]) -> dict:
    anomalies = []
    if transactions:
        amounts = [t.get("amount", 0) for t in transactions]
        mean = sum(amounts) / len(amounts) if amounts else 0
        std = (sum((a - mean) ** 2 for a in amounts) / len(amounts)) ** 0.5 if amounts else 0
        threshold = mean + 2 * std

        for t in transactions:
            if t.get("amount", 0) > threshold:
                anomalies.append({
                    "transaction_id": str(t.get("id", "")),
                    "amount": t.get("amount", 0),
                    "description": t.get("description", ""),
                    "reason": f"Valor {t.get('amount', 0):.2f} acima do threshold {threshold:.2f}",
                })

    return {
        "anomalies": anomalies,
        "mean": round(mean, 2) if transactions else 0,
        "threshold": round(threshold, 2) if transactions else 0,
    }


def _fallback_predict_budget(transactions: list[dict]) -> dict:
    categories = {}
    for t in transactions:
        cat = t.get("category", "outros")
        if cat not in categories:
            categories[cat] = 0
        categories[cat] += t.get("amount", 0)

    return {
        "budget": {cat: round(amount * 1.1, 2) for cat, amount in categories.items()},
        "total_budget": round(sum(categories.values()) * 1.1, 2),
        "message": "Orçamento sugerido com base no histórico (10% de margem).",
    }


def _fallback_tips(transactions: list[dict]) -> dict:
    return {
        "tips": [
            "Crie um fundo de emergência com 3-6 meses de despesas.",
            "Revise assinaturas e serviços recorrentes.",
            "Defina metas de economia mensais realistas.",
        ],
    }


def _fallback_chat(message: str) -> dict:
    responses = [
        "Ótima pergunta! Para melhor gerenciar seus gastos, sugiro categorizar todas as despesas.",
        "Lembre-se de revisar seu orçamento mensalmente.",
        "Uma boa prática é separar 20% da sua renda para investimentos.",
        "Tente manter seus gastos fixos abaixo de 50% da sua renda.",
    ]
    return {"response": random.choice(responses)}


FALLBACKS = {
    "classify": _fallback_classify,
    "analyze_spending": _fallback_analyze_spending,
    "anomalies": _fallback_anomalies,
    "predict_budget": _fallback_predict_budget,
    "tips": _fallback_tips,
    "chat": _fallback_chat,
}


async def classify(description: str) -> dict[str, Any]:
    messages = [
        {"role": "system", "content": SYSTEM_PROMPTS["classify"]},
        {
            "role": "user",
            "content": f"Classifique: '{description}'. Responda JSON com: category, type (receita/despesa), tags (array).",
        },
    ]

    result = await _call_ai_api(messages, temperature=0.3)
    if result:
        return result
    return _fallback_classify(description)


async def analyze_spending(transactions: list[dict]) -> dict[str, Any]:
    messages = [
        {"role": "system", "content": SYSTEM_PROMPTS["analyze_spending"]},
        {
            "role": "user",
            "content": f"Analise estes gastos: {json.dumps(transactions, ensure_ascii=False, default=str)}",
        },
    ]

    result = await _call_ai_api(messages)
    if result:
        return result
    return _fallback_analyze_spending(transactions)


async def detect_anomalies(transactions: list[dict]) -> dict[str, Any]:
    messages = [
        {"role": "system", "content": SYSTEM_PROMPTS["anomalies"]},
        {
            "role": "user",
            "content": f"Detecte anomalias: {json.dumps(transactions, ensure_ascii=False, default=str)}",
        },
    ]

    result = await _call_ai_api(messages)
    if result:
        return result
    return _fallback_anomalies(transactions)


async def predict_budget(transactions: list[dict]) -> dict[str, Any]:
    messages = [
        {"role": "system", "content": SYSTEM_PROMPTS["predict_budget"]},
        {
            "role": "user",
            "content": f"Histórico: {json.dumps(transactions, ensure_ascii=False, default=str)}",
        },
    ]

    result = await _call_ai_api(messages)
    if result:
        return result
    return _fallback_predict_budget(transactions)


async def get_tips(transactions: list[dict]) -> dict[str, Any]:
    messages = [
        {"role": "system", "content": SYSTEM_PROMPTS["tips"]},
        {
            "role": "user",
            "content": f"Dicas para: {json.dumps(transactions, ensure_ascii=False, default=str)}",
        },
    ]

    result = await _call_ai_api(messages)
    if result:
        return result
    return _fallback_tips(transactions)


async def chat(messages: list[dict]) -> dict[str, Any]:
    system_msg = {"role": "system", "content": SYSTEM_PROMPTS["chat"]}
    full_messages = [system_msg] + messages

    result = await _call_ai_api(full_messages)
    if result:
        return result

    last_user_msg = ""
    for m in reversed(messages):
        if m["role"] == "user":
            last_user_msg = m["content"]
            break

    return _fallback_chat(last_user_msg)

File: node/src/services/aiService.js
const aiApiKey = process.env.AI_API_KEY;
const aiApiUrl = process.env.AI_API_URL || 'https://api.openai.com/v1';
const aiModel = process.env.AI_MODEL || 'gpt-4';

function isConfigured() {
  return aiApiKey && aiApiKey !== 'sua-chave-aqui';
}

async function callAI(messages) {
  if (!isConfigured()) {
    return null;
  }

  const response = await fetch(`${aiApiUrl}/chat/completions`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${aiApiKey}`,
    },
    body: JSON.stringify({
      model: aiModel,
      messages,
      temperature: 0.3,
      max_tokens: 500,
    }),
  });

  if (!response.ok) {
    console.error(`AI API error: ${response.status} ${response.statusText}`);
    return null;
  }

  const data = await response.json();
  return data.choices[0]?.message?.content || null;
}

function simulateClassification(description, categories) {
  const desc = description.toLowerCase();
  const rules = {
    alimenta: ['Alimentação', 'restaurante', 'comida', 'supermercado', 'mercado', 'feira', 'lanche', 'pizza', 'ifood'],
    transporte: ['Transporte', 'uber', '99', 'taxi', 'gasolina', 'combustível', 'ônibus', 'metrô', 'pedágio', 'estacionamento'],
    moradia: ['Moradia', 'aluguel', 'condomínio', 'água', 'luz', 'energia', 'gás', 'iptu', 'manutenção'],
    saude: ['Saúde', 'médico', 'dentista', 'farmácia', 'remédio', 'plano de saúde', 'hospital', 'exame'],
    educacao: ['Educação', 'curso', 'faculdade', 'escola', 'mensalidade', 'livro', 'material'],
    lazer: ['Lazer', 'cinema', 'show', 'viagem', 'hotel', 'passagem', 'parque', 'jogo', 'streaming'],
    assinatura: ['Assinaturas', 'netflix', 'spotify', 'amazon', 'prime', 'apple'],
    salario: ['Salário', 'salário', 'pagamento', 'holerite', 'proventos'],
    investimento: ['Investimentos', 'renda fixa', 'ações', 'cdb', 'poupança', 'investimento'],
  };

  for (const [, [catName, ...keywords]] of Object.entries(rules)) {
    if (keywords.some(k => desc.includes(k))) {
      const found = categories.find(c => c.name === catName);
      if (found) return found;
    }
  }

  return null;
}

async function classifyTransaction(description, categories) {
  if (!isConfigured()) {
    return simulateClassification(description, categories);
  }

  const catList = categories.map(c => `${c.name} (${c.type})`).join(', ');

  const messages = [
    {
      role: 'system',
      content: `You are a transaction classifier. Given a description, choose the best category from this list: ${catList}. Respond only with the category name.`,
    },
    { role: 'user', content: `Classify this transaction: "${description}"` },
  ];

  const result = await callAI(messages);
  if (result) {
    const matched = categories.find(c => c.name.toLowerCase() === result.trim().toLowerCase());
    if (matched) return matched;
  }

  return simulateClassification(description, categories);
}

async function analyzeSpending(transactions, query) {
  if (!isConfigured()) {
    const total = transactions.reduce((s, t) => s + parseFloat(t.amount), 0);
    const avg = transactions.length > 0 ? total / transactions.length : 0;
    return `Análise simulada: Foram analisadas ${transactions.length} transações. Total: R$ ${total.toFixed(2)}. Média: R$ ${avg.toFixed(2)}. ${query ? 'Pergunta: ' + query : ''}`;
  }

  const summary = transactions.slice(0, 50).map(t =>
    `- ${t.description}: R$ ${parseFloat(t.amount).toFixed(2)} (${t.type})`
  ).join('\n');

  const messages = [
    {
      role: 'system',
      content: 'You are a financial analyst. Analyze the spending data and provide insights in Portuguese.',
    },
    {
      role: 'user',
      content: `Here are my transactions:\n${summary}\n\n${query ? `Question: ${query}` : 'Give me a spending analysis with insights and recommendations.'}`,
    },
  ];

  const result = await callAI(messages);
  return result || `Análise baseada em ${transactions.length} transações.`;
}

async function detectAnomalies(transactions) {
  if (transactions.length < 3) {
    return { anomalies: [], message: 'Not enough transactions to detect anomalies.' };
  }

  const amounts = transactions.map(t => parseFloat(t.amount));
  const mean = amounts.reduce((s, v) => s + v, 0) / amounts.length;
  const variance = amounts.reduce((s, v) => s + (v - mean) ** 2, 0) / amounts.length;
  const stdDev = Math.sqrt(variance);

  const anomalies = transactions.filter(t => {
    const val = parseFloat(t.amount);
    return Math.abs(val - mean) > 2 * stdDev;
  });

  if (!isConfigured()) {
    return {
      anomalies: anomalies.map(a => ({
        id: a.id,
        description: a.description,
        amount: a.amount,
        transactionDate: a.transactionDate,
        reason: `Valor ${parseFloat(a.amount).toFixed(2)} está fora de 2 desvios padrão da média ${mean.toFixed(2)}`,
      })),
      stats: { mean: mean.toFixed(2), stdDev: stdDev.toFixed(2), total: transactions.length },
    };
  }

  const anomalyList = anomalies.map(a =>
    `- ${a.description}: R$ ${parseFloat(a.amount).toFixed(2)} em ${a.transactionDate}`
  ).join('\n');

  const messages = [
    {
      role: 'system',
      content: 'You are a financial anomaly detector. Analyze the flagged transactions and explain in Portuguese why they might be anomalous.',
    },
    {
      role: 'user',
      content: `Flagged transactions:\n${anomalyList}\n\nMean: ${mean.toFixed(2)}, StdDev: ${stdDev.toFixed(2)}. Explain these anomalies.`,
    },
  ];

  const analysis = await callAI(messages);

  return {
    anomalies: anomalies.map(a => ({
      id: a.id,
      description: a.description,
      amount: a.amount,
      transactionDate: a.transactionDate,
      reason: `Valor fora do padrão (média: ${mean.toFixed(2)})`,
    })),
    stats: { mean: mean.toFixed(2), stdDev: stdDev.toFixed(2), total: transactions.length },
    analysis,
  };
}

async function predictBudget(transactions) {
  if (transactions.length === 0) {
    return { message: 'No transactions to analyze for predictions.' };
  }

  const now = new Date();
  const currentMonth = transactions.filter(t => {
    const d = new Date(t.transactionDate);
    return d.getMonth() === now.getMonth() && d.getFullYear() === now.getFullYear();
  });

  const monthlyTotals = {};
  for (const t of transactions) {
    const d = new Date(t.transactionDate);
    const key = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`;
    monthlyTotals[key] = (monthlyTotals[key] || 0) + parseFloat(t.amount);
  }

  const values = Object.values(monthlyTotals);
  const avgMonthly = values.length > 0 ? values.reduce((s, v) => s + v, 0) / values.length : 0;

  if (!isConfigured()) {
    return {
      averageMonthly: avgMonthly.toFixed(2),
      currentMonthSpending: currentMonth.reduce((s, t) => s + parseFloat(t.amount), 0).toFixed(2),
      prediction: `Previsão para o próximo mês: R$ ${avgMonthly.toFixed(2)}`,
      months: Object.entries(monthlyTotals).map(([k, v]) => ({ month: k, total: v.toFixed(2) })),
    };
  }

  const messages = [
    {
      role: 'system',
      content: 'You are a budget predictor. Analyze spending patterns and provide a prediction in Portuguese.',
    },
    {
      role: 'user',
      content: `Monthly totals: ${JSON.stringify(monthlyTotals)}. Average monthly: ${avgMonthly.toFixed(2)}. Current month: ${currentMonth.reduce((s, t) => s + parseFloat(t.amount), 0).toFixed(2)}. Predict next month.`,
    },
  ];

  const analysis = await callAI(messages);

  return {
    averageMonthly: avgMonthly.toFixed(2),
    currentMonthSpending: currentMonth.reduce((s, t) => s + parseFloat(t.amount), 0).toFixed(2),
    prediction: analysis || `Previsão: R$ ${avgMonthly.toFixed(2)}`,
    months: Object.entries(monthlyTotals).map(([k, v]) => ({ month: k, total: v.toFixed(2) })),
  };
}

async function getFinancialTips(transactions) {
  if (!isConfigured()) {
    const totalExpenses = transactions
      .filter(t => t.type === 'despesa')
      .reduce((s, t) => s + parseFloat(t.amount), 0);

    const tips = [];
    if (totalExpenses > 0) {
      tips.push('Considere criar um orçamento mensal para controlar seus gastos.');
      tips.push('Separe pelo menos 10% da sua renda para investimentos.');
      tips.push('Revise assinaturas e serviços que você não usa regularmente.');
    }
    tips.push('Mantenha uma reserva de emergência de 3 a 6 meses de despesas.');

    return { tips };
  }

  const expenses = transactions.filter(t => t.type === 'despesa').slice(0, 30);
  const expenseList = expenses.map(t =>
    `- ${t.description}: R$ ${parseFloat(t.amount).toFixed(2)} (${t.transactionDate})`
  ).join('\n');

  const messages = [
    {
      role: 'system',
      content: 'You are a financial advisor. Give personalized financial tips in Portuguese based on spending data. Be practical and specific.',
    },
    {
      role: 'user',
      content: `Based on these expenses, give me 3-5 financial tips:\n${expenseList}`,
    },
  ];

  const result = await callAI(messages);
  const tips = result ? result.split('\n').filter(l => l.trim()) : ['Mantenha um controle regular dos seus gastos.', 'Defina metas financeiras mensais.'];

  return { tips };
}

async function chat(query, context) {
  if (!isConfigured()) {
    return {
      response: `Olá! Sou o assistente financeiro. Pergunte sobre gestão de gastos, orçamento, investimentos ou economia. ${query ? `Você perguntou: "${query}"` : 'Como posso ajudar?'}`,
    };
  }

  const messages = [
    {
      role: 'system',
      content: 'You are a helpful financial assistant in Portuguese. Help users with budgeting, saving, investing, and personal finance management. Be concise and practical.',
    },
    ...(Array.isArray(context) ? context : []),
    { role: 'user', content: query },
  ];

  const result = await callAI(messages);
  return { response: result || 'Não foi possível processar sua pergunta.' };
}

module.exports = {
  classifyTransaction,
  analyzeSpending,
  detectAnomalies,
  predictBudget,
  getFinancialTips,
  chat,
};
